hive/replays/recurrent_replay.py

```diff
@@ -0,0 +1,261 @@
+        if self._max_seq_len == 2:
+            indices = indices[~done]
+        else:
+            indices = indices[~done.any(axis=1)]
+        return indices
+
+    def sample(self, batch_size):
+        """Sample transitions from the buffer. For a given transition, if it's
+        done is True, the next_observation value should not be taken to have any
+        meaning.
+
+        Args:
+            batch_size (int): Number of transitions to sample.
+        """
+        if self._num_added < self._max_seq_len + self._n_step:
+            raise ValueError("Not enough transitions added to the buffer to sample")
+        indices = self._sample_indices(batch_size)
+        batch = {}
+        batch["indices"] = indices
+        terminals = self._get_from_storage(
+            "done",
+            indices - self._max_seq_len + 1,
+            num_to_access=self._max_seq_len + self._n_step - 1,
+        )
+
+        if self._n_step == 1:
+            is_terminal = terminals
+            trajectory_lengths = np.ones(batch_size)
+        else:
+            is_terminal = terminals.any(axis=1).astype(int)
+            trajectory_lengths = (
+                np.argmax(terminals.astype(bool), axis=1) + 1
+            ) * is_terminal + self._n_step * (1 - is_terminal)
+            is_terminal = terminals[:, 1 : self._n_step - 1]
+        trajectory_lengths = trajectory_lengths.astype(np.int64)
+
+        for key in self._specs:
+            if key == "observation":
+                batch[key] = self._get_from_storage(
+                    "observation",
+                    indices - self._max_seq_len + 1,
+                    num_to_access=self._max_seq_len,
+                )
+            elif key == "action":
+                batch[key] = self._get_from_storage(
+                    "action",
+                    indices - self._max_seq_len + 1,
+                    num_to_access=self._max_seq_len,
+                )
+            elif key == "done":
+                batch["done"] = is_terminal
+            elif key == "reward":
+                rewards = self._get_from_storage(
+                    "reward",
+                    indices - self._max_seq_len + 1,
+                    num_to_access=self._max_seq_len + self._n_step - 1,
+                )
+                if self._max_seq_len + self._n_step - 1 == 1:
+                    rewards = np.expand_dims(rewards, 1)
+
+                if self._n_step == 1:
+                    rewards = rewards * np.expand_dims(self._discount, axis=0)
+
+                if self._n_step > 1:
+                    idx = np.arange(rewards.shape[1] - self._n_step + 1)[
+                        :, None
+                    ] + np.arange(self._n_step)
+
+                    # Creating a vectorized sliding window to calculate discounted returns for every element in the sequence
+                    disc_rewards = np.einsum(
+                        "ijk,k->ij", rewards[:, idx], self._discount
+                    )
```

Comment:
So this section is still a bit confusing. I had to work it out separately to make sense of what was happening. Could you write it maybe something like this?:
```
idx = np.arange(rewards.shape[1] - self._n_step + 1)[:, None] # (S-N+1) x 1
idx = idx + np.arange(self._n_step) # (S-N+1) x N
rewards = rewards[idx] # B x (S-N+1) x N
# equivalent to np.sum(rewards * self._discount[None, None, :], axis=2) 
rewards = np.einsum("ijk,k->ij", rewards, self._discount)
```
Above is just a suggestion, you can change it to make it look better. Maybe also combine first two lines. But idea is to make it clear what's happening.